documentation/writing_plugins.md

```diff
@@ -1,151 +1,492 @@
+
+There is a single plugin hook associated with Puppet library plugins:
+
+- `puppet_library`
+
+  This hook is used for plugins that make sure the Puppet library is
+  available on a target.
+
+📖 **Related information**
+
+- [Example: puppet_agent
+  plugin](https://github.com/puppetlabs/puppetlabs-puppet_agent/tree/master/tasks)
+
+## Configuring hooks
+
+The `bolt_plugin.json` not only indicates to Bolt that the module includes a
+plugin, but also allows you to map tasks to specific plugin hooks. Mapping a
+task to a plugin hook allows you to give the task any valid name you want while
+still having Bolt recognize the task as a plugin. You can also configure
+multiple hooks for a single plugin.
+
+For example, if you have a task named `my_module::my_plugin` that you want to
+use as a reference plugin, you would add the following to your
+`bolt_plugin.json` file:
+
+```json
+{
+  "hooks": {
+    "resolve_reference": {
+      "task": "my_module::my_plugin"
+    }
+  }
+}
+```
+
+Plugins can also support multiple hooks. For example, you might want to use
+a `secret_decrypt` plugin as a `resolve_reference` plugin:
 
 ```json
 {
   "hooks": {
     "resolve_reference": {
-      "task": "my_module::secret_decrypt"
+      "task": "pkcs7::secret_decrypt"
+    }
+  }
+}
+```
+
+## Plugin input
+
+Plugins can accept structured input. For example, a plugin that retrieves a list
+of targets from a service may need credentials to authenticate with the service,
+while a plugin that decrypts a value may need a path to a key pair.
+
+Because plugins are written as tasks, you can pass input to a plugin by defining
+parameters in the task metadata file and then specifying the parameters when you
+use the plugin. A simple plugin that loads a YAML file may include a single
+`filepath` parameter in the task metadata:
+
+```json
+{
+  "description": "Read YAML data from a file.",
+  "input_method": "stdin",
+  "parameters": {
+    "filepath": {
+      "type": "String",
+      "description": "The path to the YAML file."
     }
   }
 }
 ```
 
-In this case the plugin will implement two hooks: the `resolve_reference` hook
-that is explicitly defined and the `secret_decrypt` hook that is discovered from
-the task name. Bolt passes two metaparameters to all task hooks: `_config`,
-which contains the plugin configuration, and `_boltdir`, which contains the path
-to the current Boltdir.
+Depending on the `input_method` defined in the task metadata, you can access the
+plugin's input in a few different ways.
+
+- `stdin`: Read the input parameters from `stdin` and parse as JSON
+- `environment`: Read the input parameters from environment variables matching
+  the parameter's name and prefixed with `PT_`, for example `PT_filepath`
+- `powershell`: Read the input parameters from named arguments matching the
+  parameter's name
+
+Because the task metadata above uses the `stdin` input method, the task will
+need to read parameters from `stdin` and parse the input as JSON:
```

Comment:
```suggestion
Because the task metadata from the YAML plugin above uses the stdin input method, 
the task needs to read parameters from stdin and parse the input as JSON:
```